dulwich/refs.py

```diff
@@ -902,24 +902,29 @@ def set_if_equals(
             probe_ref = os.path.dirname(probe_ref)
 
         ensure_dir_exists(os.path.dirname(filename))
-        with GitFile(filename, "wb") as f:
-            if old_ref is not None:
-                try:
-                    # read again while holding the lock
-                    orig_ref = self.read_loose_ref(realname)
-                    if orig_ref is None:
-                        orig_ref = self.get_packed_refs().get(realname, ZERO_SHA)
-                    if orig_ref != old_ref:
+
+        # first, check if we need to write anything at all without
+        # acquiring the lock, which triggers a fs write, close and
+        # thus fsync()
+        if self.read_loose_ref(realname) != new_ref:
```

Comment:
Actually, what about the case where somebody else has the lock open? Previously you'd get a FileLocked error, but now the code would run fine.

Perhaps it's better to just do what we previously did but f.abort() if orig_ref == new_ref ? That way you'd still avoid the fsync() - and you'd avoid an extra read if things have changed.